virtualizarr/tests/test_readers/test_safetensors.py

```diff
@@ -0,0 +1,348 @@
+    assert vds.attrs["tensor_count"] == "1000"
+    assert vds.attrs["shape"] == "(10, 10)"
+    assert vds.attrs["dtype"] == "float32"
+
+    # Check shape of several tensors
+    for i in [0, 500, 999]:  # First, middle, and last tensor
+        tensor_name = f"tensor_{i}"
+        assert vds[tensor_name].shape == (10, 10)
+        assert vds[tensor_name].attrs["original_safetensors_dtype"] == "F32"
+
+    # Load a few tensors to check values
+    loadable_variables = [f"tensor_{i}" for i in range(3)]  # Load first 3 tensors
+    vds_with_data = open_virtual_dataset(
+        sample_safetensors_file_with_many_small_tensors,
+        loadable_variables=loadable_variables,
+    )
+
+    for i in range(3):
+        tensor_name = f"tensor_{i}"
+        tensor_data = vds_with_data[tensor_name].values
+        assert tensor_data.shape == (10, 10)
+        assert tensor_data.dtype == np.float32
+        # Check that data is not all zeros
+        assert not np.all(tensor_data == 0)
+
+
+@requires_network
+@pytest.mark.skipif(not HAS_SAFETENSORS, reason="safetensors not installed")
+def test_open_huggingface_safetensors():
+    """Test opening a SafeTensors file from Hugging Face Hub with metadata inspection only."""
+
+    # This is a widely-downloaded GPT-2 model (500mb) file from Hugging Face Hub with standard dtypes
+    vds = open_virtual_dataset(
+        "https://huggingface.co/openai-community/gpt2/resolve/main/model.safetensors"
+    )
+
+    assert len(vds.variables) > 0
+
+    # Check that some expected GPT-2 parameter tensors are present
+    # These are common parameter names in GPT-2 models
+    expected_tensors = [
+        "wte.weight",  # Word Token Embeddings
+        "wpe.weight",  # Word Position Embeddings
+        "ln_f.weight",  # Final layer norm weight
+    ]
+
+    for tensor_name in expected_tensors:
+        assert tensor_name in vds.variables
+
+    for tensor_name in expected_tensors:
+        var = vds[tensor_name]
+        assert hasattr(var, "shape")
+        assert hasattr(var, "dtype")
+        assert "original_safetensors_dtype" in var.attrs
+
+    # Check wte.weight shape - should be (vocab_size, embedding_dim)
+    wte_shape = vds["wte.weight"].shape
+    assert len(wte_shape) == 2
+    assert wte_shape[0] == 50257  # GPT-2 vocabulary size
+    assert wte_shape[1] == 768  # GPT-2 embedding dimension
+
+    # Check wpe.weight shape - should be (context_length, embedding_dim)
+    wpe_shape = vds["wpe.weight"].shape
+    assert len(wpe_shape) == 2
+    assert wpe_shape[0] == 1024  # GPT-2 context length
+    assert wpe_shape[1] == 768  # GPT-2 embedding dimension
+
+    # Test loading a small subset of data from a large tensor
+    # Only load a small slice to avoid downloading the entire 500MB file
+    loadable_vds = open_virtual_dataset(
+        "https://huggingface.co/openai-community/gpt2/model.safetensors",
+        loadable_variables=["wte.weight"],
+    )
```

Comment:
How long does this test take to run? (Anything < a few seconds is fine)

Reply:
2-3 seconds (and probably less for folks who have a faster connection than mine).